ddtrace/tracer/abandoned_spans.go

```diff
@@ -0,0 +1,190 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016 Datadog, Inc.
+
+package tracer
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+)
+
+type AbandonedList struct {
+	head *spansList
+	tail *spansList
+}
+
+type spansList struct {
+	head *spanNode
+	tail *spanNode
+	Next *spansList
+}
+
+type spanNode struct {
+	Element *span
+	Next    *spanNode
+}
+
+func (a *AbandonedList) String() string {
+	var sb strings.Builder
+	for e := a.head; e != nil; e = e.Next {
+		fmt.Fprintf(&sb, "%v", e.String())
+	}
+	return sb.String()
+}
+
+func (s *spansList) String() string {
+	var sb strings.Builder
+	for e := s.head; e != nil; e = e.Next {
+		fmt.Fprintf(&sb, "%v", e.String())
+	}
+	return sb.String()
+}
+
+func (s *spanNode) String() string {
+	sp := s.Element
+	if sp == nil {
+		return "[],"
+	}
+	return fmt.Sprintf("[Span Name: %v, Span ID: %v, Trace ID: %v],", sp.Name, sp.SpanID, sp.TraceID)
```

Comment:
I'm concerned by the length of the debug message that might come about from this. Hopefully spans like this should be relatively rare, but if there's a bug in someone's code, then there's no guarantee that it'll only be a few spans. At a minimum, we can avoid logging the same trace id multiple times for all of the spans in a trace, and instead just group them in the log line somehow.

From your local testing, can you provide an full example debug statement for this, so we can see what it looks like? I'd suggest putting that example in the PR description as well, so folks can review it.

I'd personally suggest a log line that looks something like this:
```
[[trace_id:1589394723, spans:["http.request" 513481703941, "sql.query" 1839473124, "some.name" 19823749137, "another.name" 8157910475]], [trace_id:8197401, spans:["sql.query" 390481434, "something.else" 12398472041]]]
```

It also could be helpful to report on the agent of the spans as well, for debugging purposes.

Reply:
Same question with 145

Reply:
That's a good point. The log on 145 prints each span on its own line, which would definitely become problematic if there's a bug in the user's code. The only work around that immediately comes to mind is to store these abandoned spans somewhere and print them all out at once, but I'm unsure how much better this would be overall. 

For line 53: one example of a log line might be

`[Span Name: op0, Span ID: 2363724139376108749, Trace ID: 2363724139376108749],[Span Name: op1, Span ID: 1670307605438189915, Trace ID: 1670307605438189915],[Span Name: op2, Span ID: 7640910383244335540, Trace ID: 7640910383244335540],[Span Name: op3, Span ID: 8750897367125275664, Trace ID: 8750897367125275664],[Span Name: op4, Span ID: 2480090885938891935, Trace ID: 2480090885938891935],[Span Name: op5, Span ID: 3174900340509091264, Trace ID: 3174900340509091264],[Span Name: op6, Span ID: 8226408437786668625, Trace ID: 8226408437786668625],[Span Name: op7, Span ID: 1257229564454821686, Trace ID: 1257229564454821686],[Span Name: op8, Span ID: 6104508647516361338, Trace ID: 6104508647516361338],[Span Name: op9, Span ID: 3431876901224456033, Trace ID: 3431876901224456033],`

which is definitely... very long (and could get longer). The only issue I see with grouping spans by their traces is that the structure that these spans are currently stored in doesn't group them by trace, but by Start time. So when printing, we'd have to go through and reformat the printed log appropriately. This might be fine, since this function is only being called once (when the tracer stops).

Reply:
Ahhh gotcha. I didn't realize it was bucketed by time, and not by trace. (Could you add a comment explaining that in one of these functions/struct godocs? Unless I just missed it)

Are they ordered by the oldest first, when printed? If not, I think that'd be great. It would also be good to indicate the age of the span in the log line.

Maybe we just choose a reasonable length for the log line, and truncate it if we go beyond that. It looks like 1MB is the max length for a single log line https://docs.datadoghq.com/api/latest/logs. So perhaps we should cut it off at like 900KB or something? Even that would be a massive log. So you could do something like this:

```
523 abandoned spans
[name: op0, span_id: 2363724139376108749, trace_id: 2363724139376108749, age: 1389 sec],[name: op1, span_id: 1670307605438189915, trace_id: 1670307605438189915, age: 1389 sec],[name: op2, span_id: 7640910383244335540, trace_id: 7640910383244335540, age: 1389 sec],[name: op3, span_id: 8750897367125275664, trace_id: 8750897367125275664, age: 1389 sec],[name: op4, span_id: 2480090885938891935, trace_id: 2480090885938891935, age: 1389 sec],[name: op5, span_id: 3174900340509091264, trace_id: 3174900340509091264, age: 1389 sec],[name: op6, span_id: 8226408437786668625, trace_id: 8226408437786668625, age: 1389 sec],[name: op7, span_id: 1257229564454821686, trace_id: 1257229564454821686, age: 1389 sec],[name: op8, span_id: 6104508647516361338, trace_id: 6104508647516361338, age: 1389 sec],[name: op9, span_id: 3431876901224456033, trace_id: 3431876901224456033]...
```
(with ... if beyond size limit, for example)

Reply:
Yup, can do! I agree that it'd be kinda confusing on the first (few) reads. 

They are currently ordered by start time in the list, so they should also be ordered when printed. I also considered having the logs cut off after a certain point, but I was concerned about potentially missing information when doing that. I.e., if there is a bug in a newer trace, causing many spans to be printed, but because of the log limit they don't get logged...?

I'm not sure if this is something that could happen or how many spans it would even take to get to that point 😆 Maybe a warn log like "Truncating abandoned spans..." or something would be enough to warn the user that this might happen.